src/storage.rs

```diff
@@ -0,0 +1,215 @@
+        }
+    } else {
+        Err("Could not determine root path".into())
+    }
+}
+
+pub struct Transaction<'a> {
+    tx: rusqlite::Transaction<'a>,
+}
+
+// This implementation exposes useful methods from the underlying DB transaction.
+// Note that more work would be needed for an encapsulation which doesn't leak rusqlite structs
+// (e.g. Statement or ToSql).
+impl<'a> Transaction<'a> {
+    pub fn commit(self) -> Result<()> {
+        Ok(self.tx.commit()?)
+    }
+
+    // The helper functions below are not public, to be usable only from submodules.
+    // They hide rusqlite-specific types (except for query params).
+
+    const NO_PARAMS: &'a [&'a dyn rusqlite::ToSql] = rusqlite::NO_PARAMS;
+
+    /// Execute a SQL statement taking no parameter
+    fn execute(&mut self, sql: &str) -> Result<usize> {
+        Ok(self.tx.execute(sql, Self::NO_PARAMS)?)
+    }
+
+    /// Execute a SQL statement taking unnamed parameters
+    fn execute_params(&mut self, sql: &str, params: &[&dyn rusqlite::ToSql]) -> Result<usize> {
+        Ok(self.tx.execute(sql, params)?)
+    }
+
+    /// Execute a query and create one object per returned line.
+    ///
+    /// This is similar to rusqlite::Statement::query_map_and_then(), but the passed function can
+    /// return errors that are not from rusqlite.
+    fn query_vec<T, F>(&mut self, sql: &str, f: F) -> Result<Vec<T>>
+    where
+        F: Fn(Row<'_>) -> Result<T>,
+    {
+        let mut stmt = self.tx.prepare(sql)?;
+        let mut rows = stmt.query(Self::NO_PARAMS)?;
+
+        let mut objects = Vec::new();
+        while let Some(row) = rows.next()? {
+            objects.push(f(Row::new(row))?);
+        }
+
+        Ok(objects)
+    }
+
+    fn query_single<T, F>(&mut self, sql: &str, f: F) -> Result<Option<T>>
+    where
+        F: FnOnce(Row<'_>) -> Result<T>,
+    {
+        let mut stmt = self.tx.prepare(sql)?;
+        let mut rows = stmt.query(Self::NO_PARAMS)?;
+
+        rows.next()?.map(|r| Row::new(r)).map(f).transpose()
+    }
+
+    fn count_from_table(&mut self, table_name: &str) -> Result<u64> {
+        // TODO: use a SQL parameter instead?
+        let sql = "
+SELECT count(*)
+FROM :name";
+
+        let params = rusqlite::params!(table_name);
+        let value: u32 = self.tx.query_row(&sql, params, |row| row.get(0))?;
+        Ok(value as u64)
+    }
+}
+
+/// Simple wrapper around a rusqlite::Row, mostly to avoid explicit error conversions in callbacks.
+/// It's not clear whether this is really worth it...
+struct Row<'a>(&'a rusqlite::Row<'a>);
+
+impl<'a> Row<'a> {
+    fn new(row: &'a rusqlite::Row<'a>) -> Self {
+        Self { 0: row }
+    }
+
+    fn column_count(&self) -> usize {
+        self.0.column_count()
+    }
+
+    fn get<I, T>(&self, index: I) -> Result<T>
+    where
+        I: rusqlite::RowIndex,
+        T: rusqlite::types::FromSql,
+    {
+        Ok(self.0.get(index)?)
+    }
+}
+
+// TODO: move all this to a better file
```

Comment:
It was a bit overkill for now, but I expect that in the future there will be one submodule for each table.